### PR TITLE
Convolutions

### PR DESCRIPTION
**Pull request by hunterboerner**
Todo:
- [x] FFT method "mode" switch
- [x] Fix broken test for "valid" configuration

### CODE REVIEW

File: lib/nx_signal/transforms.ex [hunk 1]
```diff
@@ -0,0 +1,23 @@
+defmodule NxSignal.Transforms do
+  import Nx.Defn
+
+  deftransform fft_nd(tensor, opts \\ []) do
+    Enum.reduce(0..(Nx.rank(tensor) - 1), tensor, fn el, acc ->
+      if el in Keyword.get(opts, :ignore, []) do
```
Comment (polvalente) [inline lib/nx_signal/transforms.ex]: I think you should receive the :axes option, and default it to [-1].

Then, you just reduce on :axes

File: lib/nx_signal/convolution.ex [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+      {a, b} when a == b ->
+        s1 = Nx.shape(volume) |> Tuple.to_list()
+        s2 = Nx.shape(kernel) |> Tuple.to_list()
+
+        # Axes initialization
+        axes = 0..(Nx.rank(volume) - 1)
+
+        axes =
+          for a <- axes, Enum.at(s1, a) != 1 || Enum.at(s2, a) == 1 do
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex]: Why is it != On one tensor and == on the other?

Comment (hunterboerner) [reply lib/nx_signal/convolution.ex]: Fixed

File: lib/nx_signal/transforms.ex (hunk 1 above)
Comment (hunterboerner) [reply lib/nx_signal/transforms.ex]: Fixed

File: lib/nx_signal/convolution.ex [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+  Convolution functions through various methods.
+
+  Follows the `scipy.signal` conventions.
+  """
+
+  import Nx.Defn
+  import NxSignal.Transforms
+
+  @doc """
+  Computes the convolution of two tensors.
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex:12]: We should add the definition of the convolution used here with Latex as well. And we can also point to the note in the Nx.conv docs.

The docstring is also missing examples with doctests

File: lib/nx_signal/convolution.ex [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    case opts[:method] do
+      :direct ->
+        direct_convolve(in1, in2, opts)
+
+      :fft ->
+        fftconvolve(in1, in2, opts)
+    end
+  end
+
+  defn correlate(in1, in2, opts \\ []) do
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex:87]: This function is missing the docs altogether. We need documentation and examples

File: lib/nx_signal/transforms.ex [hunk 5]
```diff
@@ -0,0 +1,21 @@
+defmodule NxSignal.Transforms do
+  import Nx.Defn
+
+  deftransform fft_nd(tensor, opts \\ []) do
```
Comment (polvalente) [inline lib/nx_signal/transforms.ex:5]: This function and module are missing documentation.

If we're going to keep this module private (which is fine by me), it should have @moduledoc false

Comment (polvalente) [reply lib/nx_signal/transforms.ex:5]: I believe the functions here are untested as well

File: lib/nx_signal/convolution.ex [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+  import Nx.Defn
+  import NxSignal.Transforms
+
+  @doc """
+  Computes the convolution of two tensors.
+
+  Given $f[n] \\in \\mathbb{C}^N$ and $k[n] \\in \\mathbb{C}^{K}$, we define the convolution $f * k$ by
+
+  $$
+    g(m) = (f * k)[m] = \\sum_{m=0}^{K-1} \\tilde{f}[n-m]\\tilde{k}[m],
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex]:
```suggestion
    g[n] = (f * k)[n] = \\sum_{m=0}^{K-1} f[n-m]k[m],
```

File: lib/nx_signal/convolution.ex [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+  Follows the `scipy.signal` conventions.
+  """
+
+  import Nx.Defn
+  import NxSignal.Transforms
+
+  @doc """
+  Computes the convolution of two tensors.
+
+  Given $f[n] \\in \\mathbb{C}^N$ and $k[n] \\in \\mathbb{C}^{K}$, we define the convolution $f * k$ by
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex]:
```suggestion
  Given $f[n]$ of length $N$ and $k[n]$ of length ${K}$, we define the convolution $(f * k)[n]$ by
```

File: lib/nx_signal/convolution.ex [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+  $$
+    g(m) = (f * k)[m] = \\sum_{m=0}^{K-1} \\tilde{f}[n-m]\\tilde{k}[m],
+  $$
+
+  where
+
+  $$
+  \\tilde{f}[n] =
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex:23]:
```suggestion
  f[n] =
```

File: lib/nx_signal/convolution.ex [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+  \\begin{cases}
+    0 & n < 0 \\\\\\\\
+    0 & n \\geq N \\\\\\\\
+    f[n] & \\text{otherwise}
+  \\end{cases}
+  $$
+
+  and $\\tilde{k}$ is defined similarly.
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex:31]:
```suggestion
  $k[n]$ is defined similarly for $n < 0$ and $n >= K$.
  $g[n]$ has length $K + N - 1$ when `mode: :full`.
```

Comment (polvalente) [inline lib/nx_signal/convolution.ex]:
```suggestion
  where $f[n]$ and $k[n]$ are assumed to be zero outside of their definition boundaries.
  
  $g[n]$ has length $N + K - 1$ when `mode: :full`
```

File: lib/nx_signal/convolution.ex [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+   where
+
+   $$
+   \\tilde{f}[n] =
+   \\begin{cases}
+     0 & n < 0 \\\\\\\\
+     0 & n \\geq N \\\\\\\\
+     f[n] & \\text{otherwise}
+   \\end{cases}
+   $$
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex]:
```suggestion
   $$
     g[n] = (f * k)[n] = \\sum_{m = 0}^{K - 1}f[n - m]\\overline{k[K - 1 - m]}
   $$
```

File: lib/nx_signal/convolution.ex [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+    * `:method` - One of `:fft` or `:direct`. Defaults to `:direct`.
+    * `:mode` - One of `:full`, `:valid`, or `:same`. Defaults to `:full`.
+
+  ## Examples
+
+    iex> NxSignal.Convolution.correlate(Nx.tensor([1,2,3]), Nx.tensor([3,4,5]))
+    #Nx.Tensor<
+      c64[5]
+      [5.0-0.0i, 14.0-0.0i, 26.0-0.0i, 18.0-0.0i, 9.0-0.0i]
```
Comment (polvalente) [inline lib/nx_signal/convolution.ex]: We shouldn't upcast real values to complex

Comment (hunterboerner) [reply lib/nx_signal/convolution.ex]: Done

**Pull request merged**